### DISCUSSION TITLE
postCreateCommand in image

### DISCUSSION THREAD

**Discussion by Chuxel** [devcontainers/community#5] [Q&A] [answered]
From @stephenandary 

Hello :). I am wondering about a use case where I can build and push a devcontainer including post create commands so that a developer just needs to reference the built image. Some post create commands take very long and it would be nice to just build when updating for people. I tried to do this but when I referenced the image, it ran all of the build again. Wondering if doing this is possible (maybe I did something wrong).

Github Action Workflow Used: https://github.com/stephenandary/Dev_Containers/blob/main/.github/workflows/Rust-Dev_image_release.yml#L1-L30

**Comment by Chuxel** [comment, marked as answer]
postCreateCommand happens after the container has been created. As a result, its contents cannot be added to an image.  Its intended to perform steps that requires the source code to be present to execute. A classic example is `yarn install`.

However, [if this is your postCreateCommand](https://github.com/stephenandary/Dev_Containers/blob/main/Rust/Rust-Dev/.devcontainer/scripts/postCreateCommand.sh), then it looks like all of those contents can be executed w/o the source code, so there's two options.

## Add to a Dockerfile

You can add many of these commands to your Dockerfile. You can add it as follows:
```
COPY ./scripts/postCreateCommand.sh /tmp
RUN bash /tmp/postCreateCommand.sh && rm -f /tmp/postCreateCommand.sh
```
Or to run as a user other than `root`:
```
COPY ./scripts/postCreateCommand.sh /tmp
USER vscode
RUN bash /tmp/postCreateCommand.sh && rm -f /tmp/postCreateCommand.sh
```

A gotcha is that sometimes things require a `login` or  `interactive` shell to work properly.  Just add `-li` after bash to get one. You'll get a warning you can ignore about input devices, but otherwise it's fine to do.

```
COPY ./scripts/postCreateCommand.sh /tmp
USER vscode
RUN bash -li /tmp/postCreateCommand.sh && rm -f /tmp/postCreateCommand.sh
```

## Use a local feature

However, if you want to run commands that depend on a specific Dev Container Feature, you can use a **local feature**. All [Dev Container Features capabilities](https://containers.dev/implementors/features/) are available in local features, you just don't have to publish them if you don't want to.  Here's how to convert that script into a local feature:

1. Create `.devcontainer/local-feature`

2. Add `devcontainer-feature.json` into the folder with the following contents:
    ```json
    {
        "id": "local-feature",
        "version": "0.0.1",
        "installsAfter": ["ghcr.io/devcontainers/features/rust"]
    }
   ```
   The `installsAfter` is because the script uses cargo. If preferred you can just override the install order in devcontainer.json.
   
3. Now if you can run everything as `root`, you could just rename `postCreateCommand.sh` to `install.sh` and place it in `.devcontainer/local-feature`.  

    **[OPTIONAL]** However, if you need to run as the non-root user (`remoteUser`) like `postCreateCommand` does, you can take advantage of the `_REMOTE_USER` environment variable with `su`. For example:
    1. Move `postCreateCommand.sh` to .devcontainer/local-feature
    2. Create `install.sh` in the same spot with the following contents: 
        ```bash
        #!/bin/bash
        chmod 777 ./postCreateCommand.sh
        su ${_REMOTE_USER} -c 'bash postCreateCommand.sh'
        ```
        You can also use `bash -li` here as needed.
4. Add the reference under `features` in `devcontainer.json`:
    ```jsonc
    {
        //...
        "features": {
		//..
		"./local-feature": {}
	},
        //...
    }
   ```

That help?

**Reply by andar1an** [reply]
This is amazing, thank you so much for the response. I ran the script as a post-create command due to ownership issues, and it would have been a lot of work to figure out where I needed to take ownership of directories as not root post-build. So I will be able to avoid ownership issues for non-root by defining the users prior to running the script?! That is so cool, thank you so much again!

**Reply by Chuxel** [reply]
Glad it helped! @bamurtaugh raised a tracking issue for the kinds of "how-tos" should get added to containers.devs site (and probably the spec repo), so thanks for raising it! https://github.com/devcontainers/spec/issues/137

**Reply by andar1an** [reply]
Very glad to learn about that site. In my experience with dev containers, I usually reference microsoft docs, and the containers are not always the most efficient, or concise. This can make them read a bit hacky. It will be really nice to see some how-to's with good examples! Really excited to try out what you mentioned above this weekend with my gh action. Im a big believer in giving people a consistent build environment with minimal set up required for each project. This project is awesome for potentially managing all of my different devcontainer builds from a single repo! <3

**Comment by andar1an** [comment]
@Chuxel, within `devcontainer-feature.json`, how do you put multiple features? I tried: 
```
{
    "id": "local-feature-1",
    "version": "0.0.1",
    "installsAfter": ["ghcr.io/devcontainers/features/something"]
},
{
    "id": "local-feature-2",
    "version": "0.0.1",
    "installsAfter": ["ghcr.io/devcontainers/features/somethingelse"]
}
```

and variations on:

```
{
    "id": "local-feature-1",
    "version": "0.0.1",
    "installsAfter": ["ghcr.io/devcontainers/features/something"],
    
    "id": "local-feature-2",
    "version": "0.0.1",
    "installsAfter": ["ghcr.io/devcontainers/features/somethingelse"]
}
```

**Reply by Chuxel** [reply]
Each `devcontainer-feature.json` is for a single feature.  However, you can have multiple features in one repository that are then published as a group. This is how https://github.com/devcontainers/feature-starter is set up with a `src` folder that has multuple features under it.

The same general approach works for local features as well. You just use different sub-folders.

**Reply by andar1an** [reply]
I think https://github.com/devcontainers/feature-starter may have answered. Do you do 

```
/local-features
    |__ /feature1
        |__ devcontainer-feature.json
        |__ install.sh
    |__ /feature2
        |__ devcontainer-feature.json
        |__ install.sh
```

**Reply by andar1an** [reply]
@Chuxel Perfect timing haha

**Reply by andar1an** [reply]
I am so excited about all the repos I didn't see before. The features Repo SERIOUSLY simplified my Dockerfile!

**Reply by Chuxel** [reply]
Yep! Though local features need to be in a sub-folder either in `.devcontainer` or a sub folder to `.devcontainer.json`.

**Reply by andar1an** [reply]
When I use local features like so: 
```
#!/bin/bash
su ${_REMOTE_USER} -c 'bash postCreateCommand.sh'
```
I get the equivalent of  `bash: postCreateCommand.sh: Permission denied`

Any ideas about what may be causing this? I am having trouble determining the cause. 

I feel like it should be using the vscode remote user defined in devcontainer.json to install the features, but does that mean I don't have access to scripts because they are owned by root? 

<https://github.com/jamminmusic/Containers/tree/main/devcontainers/rust/.devcontainer>

**Reply by andar1an** [reply]
@Chuxel As a note the reason I needed to install rust as vscode user was due to the following error if not using devcontainer as root:
`warning: spurious network error (2 tries remaining): failed to create temporary file '/usr/local/cargo/registry/index/github.com-1ecc6299db9ec823/.git/objects/pack/pack_git2_80e6e668392a3819': Permission denied; class=Os (2)`

Is it possible to install a feature as remote user?

**Reply by Chuxel** [reply]
For the first issue, you may need to `chown` first since I think root is the owner, and permissions are 700. So only root can access it.  

`chmod o+r your-filename-here`

Also, FYI, there's a rust Feature now too (`ghcr.io/devcontainers/features/rust`) if you want to see how it works: https://github.com/devcontainers/features/tree/main/src/rust 

@chrmarti @joshspicer I know this was set to 700 for caching purposes, but we could set this to be readable by everyone. Not sure there's value in being this restrictive.  Thoughts?

**Reply by andar1an** [reply]
Thanks so much. I am trying to use the Rust feature. The reason I have avoided it in the past is the same issue I am experiencing now. It has been much easier for me to just install Rust. I moved the Rust question to https://github.com/orgs/devcontainers/discussions/13 so I don't overload this.

**Reply by andar1an** [reply]
@Chuxel I have tried taking ownership and changing file permissions at least 10 different ways. Still not able to reference the file from install.sh

**Reply by andar1an** [reply]
This seemed to do it:
```
chmod 777 -R $PWD
su ${_REMOTE_USER} -c './filename.sh'
```
